rmw_connext_cpp/src/functions.cpp

```diff
@@ -1023,6 +1037,71 @@ rmw_wait(rmw_subscriptions_t * subscriptions,
            wait_timeout);
 }
 
+bool
+__process_service_name(
+  const char * service_name,
+  bool avoid_ros_namespace_conventions,
+  char ** request_partition_str,
+  char ** response_partition_str,
+  char ** service_str)
+{
+  bool success = true;
+  rcutils_string_array_t name_tokens = rcutils_get_zero_initialized_string_array();
+
+  // get actual data subsription object
+  // allocates memory, but doesn't have to be freed.
+  // partition operater takes ownership of it
+  name_tokens = rcutils_split_last(service_name, '/');
+  if (name_tokens.size == 1) {
+    if (!avoid_ros_namespace_conventions) {
+      *request_partition_str = DDS_String_dup(ros_service_requester_prefix);
+      *response_partition_str = DDS_String_dup(ros_service_response_prefix);
+    }
+    *service_str = DDS_String_dup(name_tokens.data[0]);
+  } else if (name_tokens.size == 2) {
+    if (avoid_ros_namespace_conventions) {
+      // no ros_service_*_prefix, so store the user's namespace directly
+      *request_partition_str = DDS_String_dup(name_tokens.data[0]);
+      *response_partition_str = DDS_String_dup(name_tokens.data[0]);
+    } else {
+      // concat the ros_service_*_prefix with the user's namespace
+      rcutils_allocator_t allocator = rcutils_get_default_allocator();
+      char * request_concat_str = rcutils_format_string(
+        allocator,
+        "%s/%s", ros_service_requester_prefix, name_tokens.data[0]);
+      if (!request_concat_str) {
+        RMW_SET_ERROR_MSG("could not allocate memory for partition string")
+        success = false;
+        goto end;
+      }
+      char * response_concat_str = rcutils_format_string(
+        allocator,
+        "%s/%s", ros_service_response_prefix, name_tokens.data[0]);
+      if (!response_concat_str) {
+        allocator.deallocate(request_concat_str, allocator.state);
```

Comment:
No, the `request_concat_str` points to allocated memory at this point. If the deallocate did not happen, then it would be leaked here. There is nothing that needs to be deallocated in either of the other two cases, `request_concat_str` or `concat_str`.